### PR TITLE
Fix issue #186

### PR DESCRIPTION
**Pull request by endison1986**

### CODE REVIEW

File: dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java [hunk 1]
```diff
@@ -79,15 +78,21 @@ public int addObject(Object newClass) {
...
+            } else {
+                // whether the existing index and the new index are the same. If they are different, it means there is a hash conflict.
+                if(idx != unsafe.getIntVolatile(null, i)) {
+                    useFallbackMap.set(true);
+                }
+                return idx;
+            }
         } else {
-            if (!controlMap.containsKey(newClass)) {
+            if (cache.getVolatile(currentIndex - 1) != newClass) {
```
Comment (enricostara) [inline dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java]: like above

File: dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java [hunk 2]
```diff
@@ -105,23 +110,31 @@ public int getObjectIndex(Object klass) {
...
     }
 
     public int getObjectIndexVolatile(Object klass) {
         if (useFallbackMap.get()) {
             return fallbackMap.get((Class<?>) klass);
         }
         int identityHashCode = capHashCode(System.identityHashCode(klass), hashBit);
-        return unsafe.getIntVolatile(null, getIdentityAddress(identityHashCode, memoryAddress));
+        int index = unsafe.getIntVolatile(null, getIdentityAddress(identityHashCode, memoryAddress));
+        if (index != 0 && cache.get(index - 1) == klass) {
```
Comment (enricostara) [inline dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java]: like above

File: dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java [hunk 3]
```diff
@@ -105,23 +110,31 @@ public int getObjectIndex(Object klass) {
             return fallbackMap.get((Class<?>) klass);
         }
         int identityHashCode = capHashCode(System.identityHashCode(klass), hashBit);
-        return unsafe.getInt(getIdentityAddress(identityHashCode, memoryAddress));
+        int index = unsafe.getInt(getIdentityAddress(identityHashCode, memoryAddress));
+        if (index != 0 && cache.get(index - 1) == klass) {
+            return index;
```
Comment (enricostara) [inline dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java:115]: like above

File: dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java [hunk 4]
```diff
@@ -79,15 +78,21 @@ public int addObject(Object newClass) {
...
-                    fallbackMap.get((Class<?>) newClass) :
-                    atomicIndex.incrementAndGet();
-            unsafe.putIntVolatile(null, i, idx);
-            controlMap.put(newClass, idx);
-            return idx;
+        if(currentIndex == 0) {
+            int idx = fallbackMapEnabled ? fallbackMap.get((Class<?>) newClass) : atomicIndex.incrementAndGet();
+            // use cas to check, is there an index to an existing class
+            if(unsafe.compareAndSwapInt(null, i, 0, idx)) {
+                cache.set(idx - 1, newClass);
```
Comment (enricostara) [inline dominion-ecs-engine/src/main/java/dev/dominion/ecs/engine/system/ClassIndex.java]: We cache currentIndex without -1, we don't care about wasting a position if it helps readability

**Review by enricostara — CHANGES_REQUESTED** [review]
A minor fix requested in ClassIndex

**Comment by endison1986** [comment]
sorry，I don’t quite understand the operation of review，If there is anything incorrect, please correct me 🥺
I add LIC header and optimize readability with new commit.

**Review by enricostara — APPROVED** [review]

**Pull request merged**